internal/controller/accessrequest_controller.go

```diff
@@ -33,30 +38,241 @@ type AccessRequestReconciler struct {
+	if err != nil {
+		return "", fmt.Errorf("error updating access request status to granted: %w", err)
+	}
+	return api.GrantedStatus, nil
+}
+
+func (r *AccessRequestReconciler) updateStatusWithRetry(ctx context.Context, ar *api.AccessRequest, status api.Status, details string) error {
+	return retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		err := r.Get(ctx, client.ObjectKeyFromObject(ar), ar)
+		if err != nil {
+			return err
+		}
+		ar.UpdateStatus(status, details)
+		return r.Status().Update(ctx, ar)
+	})
+}
+
+// TODO
+func (r *AccessRequestReconciler) updateAppProjectRBAC(ctx context.Context, ar *api.AccessRequest) error {
+	// 1. retrieve the Application
+	// 2. verify if CR is approved
+	// 3. retrieve the AppProject
+	// 4. assign user in the desired role in the AppProject
+	return nil
+}
+
+type AllowedResponse struct {
+	Allowed bool
+	Message string
+}
+
+// TODO
+func (r *AccessRequestReconciler) Allowed(ctx context.Context, ar *api.AccessRequest) (AllowedResponse, error) {
+	return AllowedResponse{}, nil
+}
+
+// handleAccessExpired will verify if the given AccessRequest is expired. If
+// so, it will remove the Argo CD access for the subject and update the
+// AccessRequest status field.
+func (r *AccessRequestReconciler) handleAccessExpired(ctx context.Context, ar *api.AccessRequest) (bool, error) {
+	expired := false
+	if ar.Status.ExpiresAt != nil &&
+		ar.Status.ExpiresAt.Time.After(time.Now()) {
+
+		err := r.removeArgoCDAccess(ar)
+		if err != nil {
+			return expired, fmt.Errorf("error removing access for expired request: %w", err)
+		}
+		err = r.updateStatusWithRetry(ctx, ar, api.ExpiredStatus, "")
+		if err != nil {
+			return expired, fmt.Errorf("error updating access request status to expired: %w", err)
+		}
+		expired = true
+	}
+	return expired, nil
+}
+
+// handleFinalizer will check if the AccessRequest is being deleted and
+// proceed with the necessary clean up logic if so. If the object is not
+// being deleted, it will register the AccessRequest finalizer in the live
+// state. The function will return a boolean value to determine if the object
+// was deleted.
+func (r *AccessRequestReconciler) handleFinalizer(ctx context.Context, ar *api.AccessRequest) (bool, error) {
+
+	deleted := false
+	// examine DeletionTimestamp to determine if object is under deletion
+	if ar.ObjectMeta.DeletionTimestamp.IsZero() {
+		// The object is not being deleted, so if it does not have the
+		// finalizer, then we register it.
+		if !controllerutil.ContainsFinalizer(ar, AccessRequestFinalizerName) {
+			err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+				err := r.Get(ctx, client.ObjectKeyFromObject(ar), ar)
+				if err != nil {
+					return err
+				}
+				controllerutil.AddFinalizer(ar, AccessRequestFinalizerName)
+				return r.Update(ctx, ar)
+
+			})
+			if err != nil {
+				return deleted, fmt.Errorf("error adding finalizer: %w", err)
+			}
+		}
+	} else {
```

Comment:
Instead of deep nesting, I'd return early for the "not deleting" case and unwrap this. Then you can do away with the `deleted` var and instead just return hard-coded booleans.

Reply:
Nice suggestion. 
Modified.